utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala

```diff
@@ -0,0 +1,185 @@
+  val array =
+    if (sorted) arr
+    else arr.sortBy(_._1)
+
+  def length = array.length
+  def midpoint = pow2ceil()
+
+  @tailrec private def pow2ceil(i: Int = 1): Int = {
+    if (2 * i >= length) {
+      i
+    } else {
+      pow2ceil(2 * i)
+    }
+  }
+
+  @tailrec private def binarySearch(rr: K,
+                                    idx: Int = 0,
+                                    step: Int = midpoint): Option[Int] = {
+    if (array.length == 0) {
+      None
+    } else if (rr.overlaps(array(idx)._1)) {
+      Some(idx)
+    } else if (step == 0) {
+      None
+    } else {
+      val stepIdx = idx + step
+      val nextIdx = if (stepIdx >= length ||
+        (!rr.overlaps(array(stepIdx)._1) &&
+          rr.compareTo(array(stepIdx)._1) < 0)) {
+        idx
+      } else {
+        stepIdx
+      }
+      binarySearch(rr, nextIdx, step / 2)
+    }
+  }
+
+  @tailrec private def expand(rr: K,
+                              idx: Int,
+                              step: Int,
+                              list: List[(K, T)] = List.empty): List[(K, T)] = {
+    if (idx < 0 ||
+      idx >= length ||
+      !rr.overlaps(array(idx)._1)) {
+      list
+    } else {
+      expand(rr, idx + step, step, array(idx) :: list)
+    }
+  }
+
+  /**
+   * Insert Iterator of (K,V) items into existing RangeSearchableArray
+   * @param kvs (K,V) tuples to insert into RangeSearchableArray
+   * @return new RangeSearchableArray with inserted values
+   */
+  def insert(kvs: Iterator[(K, T)]): RangeSearchableArray[K, T] = {
+    RangeSearchableArray((array ++ kvs).sortBy(_._1))
+  }
+
+  /**
+   * Filters items in RangeSearchableArray based on predicate on (K,V) tuples
+   * @param pred predicate to filter elements by
+   * @return new RangeSearchableArray with filtered elements
+   */
+  def filter(pred: ((K, T)) => Boolean): RangeSearchableArray[K, T] = {
+    RangeSearchableArray(array.filter(r => pred(r._1, r._2)))
+  }
+
+  /**
+   * Maps values from T to T2
+   * @param f Function mapping T to T2
+   * @tparam T2 new type to map values to
+   * @return new RangeSearchableArray with mapped values
+   */
+  def mapValues[T2: ClassTag](f: T => T2): RangeSearchableArray[K, T2] = {
+    RangeSearchableArray(array.map(r => (r._1, f(r._2))))
+  }
+
+  /**
+   * Filters elements in array by overlapping Interval
+   * @param rr Interval to filter by
+   * @return Iterable of elements filtered by Interval rr
+   */
+  def get(rr: K): Iterable[(K, T)] = {
+
+    val optIdx = binarySearch(rr)
+
+    optIdx.toIterable
+      .flatMap(idx => {
+        expand(rr, idx, -1) ::: expand(rr, idx + 1, 1)
+      })
+  }
+}
+
+class RangeSearchableArraySerializer[K <: Interval[K]: ClassTag, T: ClassTag, TS <: Serializer[T], KS <: Serializer[K]](
+    private val kSerializer: KS,
+    private val tSerializer: TS) extends Serializer[RangeSearchableArray[K, T]] {
+
+  def tTag: ClassTag[T] = implicitly[ClassTag[T]]
+  def kTag: ClassTag[K] = implicitly[ClassTag[K]]
```

Comment:
@fnothaft let me know if this is what you meant for discovering types at runtime. Is there another change that should be made here?